src/Streamly/Internal/Data/Fold/Concurrent/Channel/Type.hs

```diff
@@ -0,0 +1,330 @@
+-- |
+-- Module      : Streamly.Internal.Data.Fold.Concurrent.Channel.Type
+-- Copyright   : (c) 2017, 2022 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+module Streamly.Internal.Data.Fold.Concurrent.Channel.Type
+    ( Channel (..)
+    , newChannel
+    , Config
+    , sendToWorker
+    , checkFoldStatus
+    , dumpSVar
+    )
+where
+
+#include "inline.hs"
+
+import Control.Concurrent (ThreadId, myThreadId, tryPutMVar)
+import Control.Concurrent.MVar (MVar, newEmptyMVar, takeMVar)
+import Control.Exception (SomeException(..))
+import Control.Monad (void)
+import Control.Monad.Catch (throwM)
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Data.IORef (IORef, newIORef, readIORef)
+import Data.List (intersperse)
+import Streamly.Internal.Control.Concurrent (MonadAsync, askRunInIO)
+import Streamly.Internal.Control.ForkLifted (doFork)
+import Streamly.Internal.Data.Fold (Fold(..))
+import Streamly.Internal.Data.Stream.Channel.Dispatcher (dumpSVarStats)
+import Streamly.Internal.Data.Stream.Channel.Worker (sendWithDoorBell)
+
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.Stream.StreamD.Type as D
+
+import Streamly.Internal.Data.Stream.Channel.Types
+
+data Channel m a b = Channel
+    {
+    -- FORWARD FLOW: Flow of data from the driver to the consuming fold
+
+    -- XXX This is inputQueue instead.
```

Comment:
Rename the fields of the channel accordingly?
Or open an issue about it foe doing it later?